src/components/Loans/OpenLoans/components/BulkClaimReturnedModal/BulkClaimReturnedModal.js

```diff
@@ -0,0 +1,267 @@
+    requestCounts: PropTypes.object.isRequired,
+  }
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      additionalInfo: '',     // The 'additional information' text included as part of a claim returned POST
+      operationResults: [],   // An array of results returned from the server after a POST
+      operationState: 'pre',  // Whether we're at the start ('pre') or end ('post') of a transaction
+    };
+  }
+
+  // Return the number of requests for item @id
+  // requestCounts is an object of the form {<item id>: <number of requests>},
+  // with keys present only if the item in question has >0 requests. Thus, if
+  // none of the selected items have any requests on them, requestCounts === {}.
+  getRequestCountForItem = id => this.props.requestCounts[id] || 0;
+
+  handleAdditionalInfoChange = e => {
+    this.setState({ additionalInfo: e.target.value });
+  };
+
+  claimItemReturned = (loan) => {
+    if (!loan) return null;
+
+    this.props.mutator.loanId.replace(loan.id);
+    return this.props.mutator.claimReturned.POST(
+      {
+        itemClaimedReturnedDateTime: moment().format(),
+        comment: this.state.additionalInfo,
+      }
+    );
+  }
+
+  claimAllReturned = () => {
+    this.setState({ operationResults: [] });
+    const promises = Object
+      .values(this.props.checkedLoansIndex)
+      .map(loan => this.claimItemReturned(loan).catch(e => e));
+    Promise.all(promises)
+      .then(results => this.finishClaims(results));
+  }
+
+  finishClaims = (results) => {
+    this.setState({
+      operationResults: results,
+      operationState: 'post',
+    });
+  }
+
+  // @results: array of results of the claim-returned operation for each loan
+  // Each item in the array will either be a simple object that represents a
+  // successful change (including an id value) or a larger object representing
+  // a CORS error (identifiable by the properties ok: false and url).
+  //
+  // @return array of loan IDs for which the operation returned an error.
+  getUnchangedLoansList = (results) => {
+    return results.filter(r => r.ok === false).map(r => r.url?.match(/^.*\/(.*)\/claim-item-returned/)[1]);
+  }
+
+  onCancel = () => {
+    this.setState({ operationState: 'pre' });
+    this.props.onCancel();
+  }
+
+  // Renders a modal dialog containing a list of items to be claimed returned for both
+  // pre-operation review and post-operation feedback (i.e., there's one modal, but the
+  // contents will be different before and after clicking the 'Confirm' button.)
+  render() {
+    const {
+      checkedLoansIndex,
+      intl,
+      open
+    } = this.props;
+    const {
+      additionalInfo,
+      operationResults,
+      operationState,
+    } = this.state;
+    const loans = checkedLoansIndex ? Object.values(checkedLoansIndex) : [];
+    // List of loans for which the claim returned operation failed
+    const unchangedLoans = this.getUnchangedLoansList(operationResults);
```

Comment:
@Baroquem I wonder if we could use `this.getUnchangedLoansList` right in `finishClaims` that way it would execute only once and not during every `render`. What do you think about that?

Reply:
I think it's a great idea!